### PR TITLE
[cli] indicates whether @vercel/speed-insights or @vercel/analytics are used

### PR DESCRIPTION
**Pull request by feugy**
### 🧐 What's in there?

When deploying to Vercel, the build container needs to know whether the deployed apps is using `@vercel/speed-insights` and `@vercel/analytics` or not.

As discussed on slack, using `builds.json` is a nice way to convey this piece of information.

### CODE REVIEW

File: packages/cli/test/fixtures/unit/commands/git/connect/unlinked/git/HEAD
Comment (Ethan-Arrowood) [inline packages/cli/test/fixtures/unit/commands/git/connect/unlinked/git/HEAD:1]: why are these removed in this PR?

**Review by Ethan-Arrowood — COMMENTED** [review]
lgtm other than than the couple fixture files that are deleted. Not sure how that's related

**Comment by Ethan-Arrowood** [comment]
don't forget to add a Changeset with `npx changeset`

**Review by styfle — CHANGES_REQUESTED** [review]
Please restore the deleted test fixture 

```
packages/cli/test/fixtures/unit/commands/git/connect/unlinked
```

File: packages/cli/test/integration-2.test.ts [hunk 1]
```diff
@@ -1165,6 +1165,11 @@ test('[vc build] should build project with `@vercel/speed-insights`', async () =
     expect(output.stderr).toContain(
       'The `VERCEL_ANALYTICS_ID` environment variable is deprecated and will be removed in a future release. Please remove it from your environment variables'
     );
+    console.log(output.stderr, output.stdout);
```
Comment (EndangeredMassa) [inline packages/cli/test/integration-2.test.ts]:
```suggestion
```

File: packages/cli/test/fixtures/unit/commands/git/connect/unlinked/git/HEAD
Comment (feugy) [reply packages/cli/test/fixtures/unit/commands/git/connect/unlinked/git/HEAD:1]: Apologies, that was no intentional indeed

File: packages/cli/test/helpers/prepare.js [hunk 2]
```diff
@@ -433,6 +433,24 @@ module.exports = async function prepare(session, binaryPath, tmpFixturesDir) {
...
+          framework: null,
+          installCommand: 'echo "skipping install"',
+        },
+      }),
+      'package.json': JSON.stringify({
+        scripts: {
+          build: 'mkdir -p public && echo hi > public/index.txt',
+        },
+        dependencies: {
+          '@vercel/web-analytics': '0.0.1',
```
Comment (tobiaslins) [inline packages/cli/test/helpers/prepare.js]: the dependency here is wrong. its `@vercel/analytics`

File: packages/cli/src/commands/build/index.ts [hunk 3]
```diff
@@ -807,3 +811,21 @@ function mergeFlags(
...
+}
+
+function isUsingWebAnaytics(pkg: PackageJson | null): boolean {
+  const dependencies = [
+    ...Object.keys(pkg?.dependencies ?? {}),
+    ...Object.keys(pkg?.devDependencies ?? {}),
+  ];
+
+  return dependencies.some(d => d === '@vercel/analytics');
+}
```
Comment (tobiaslins) [inline packages/cli/src/commands/build/index.ts]: maybe we can abstract away checking if a package is installed
```suggestion
function isPackageInstalled(name: string; pkg: PackageJson | null) {
  const dependencies = [
    ...Object.keys(pkg?.dependencies ?? {}),
    ...Object.keys(pkg?.devDependencies ?? {}),
  ];

  return dependencies.some(d => d === name);
}

function isUsingSpeedInsights(pkg: PackageJson | null): boolean {
  return isPackageInstalled('@vercel/speed-insights', pkg);
}

function isUsingWebAnaytics(pkg: PackageJson | null): boolean {
   return isPackageInstalled('@vercel/analytics', pkg);
}
```

**Review by tobiaslins — DISMISSED** [review]
Overall great! the small proposed refactor would be nice!

**Review by tobiaslins — DISMISSED** [review]

**Review by Ethan-Arrowood — DISMISSED** [review]

File: packages/cli/src/commands/build/index.ts [hunk 4]
```diff
@@ -807,3 +811,20 @@ function mergeFlags(
     return [];
   });
 }
+
+function isPackageInstalled(name: string, pkg: PackageJson | null): boolean {
+  const dependencies = [
+    ...Object.keys(pkg?.dependencies ?? {}),
+    ...Object.keys(pkg?.devDependencies ?? {}),
+  ];
```
Comment (styfle) [inline packages/cli/src/commands/build/index.ts]: Right now these packages are iterated twice because `isPackageInstalled()` is always called twice. 

Let's avoid that by hoisting out of this function (thats basically how it was working before)

You could even use a Set to avoid `.some()` iteration again.

Comment (styfle) [reply packages/cli/src/commands/build/index.ts]: For example:

```js
const dependencies = new Set();
for (const key in pkg?.dependencies ?? {}) {
  dependencies.add(key);
}
for (const key in pkg?.devDependencies ?? {}) {
  dependencies.add(key);
}


if (dependencies.has('@vercel/speed-insights')) {
  // do stuff
}

if (dependencies.has('@vercel/analytics')) {
  // do stuff
}
```

Comment (feugy) [reply packages/cli/src/commands/build/index.ts]: Good call. @chriswdmr suggested that we convey the package versions instead of a boolean.
I made it a map, and it's built only once

Comment (feugy) [reply packages/cli/src/commands/build/index.ts]: In the unlikely event of the same package being in both `devDependencies` and `dependencies`, the later will take over.

**Review by styfle — APPROVED** [review]

File: packages/cli/src/commands/build/index.ts [hunk 5]
```diff
@@ -106,6 +106,7 @@ export interface BuildsManifest {
   argv: string[];
   error?: any;
   builds?: SerializedBuilder[];
+  features?: { speedInsightsVersion: string; webAnalyticsVersion: string };
```
Comment (cb1kenobi) [inline packages/cli/src/commands/build/index.ts:109]: @feugy FYI, this line is failing type checking: https://github.com/vercel/vercel/actions/runs/6421634817/job/17436437058?pr=10667#step:9:57

Fixed in my PR: https://github.com/vercel/vercel/pull/10667/files#diff-452a115a951a23488dd3257a83934cc9439918fbd41143a0ded219164dd13c6f
